HDDS-7277. [Snapshot] Implement snapshot key listing in Ozone Shell

xBis7 opened the pull request:
## What changes were proposed in this pull request?

This error was coming from `OzoneAddress.ensureBucketAddress()` but if we were to modify that method to accept the snapshot name as part of the keyName while checking a bucket path, then every bucket command would accept a snapshot path appended to it. Therefore, a new method was added under `OzoneAddress` called `ensureSnapshotAddress()`. This method will accept both a bucket and snapshot but it will only be used with the `ozone sh key list` command so that there won't be any unpredictable behavior from other commands.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7277

## How was this patch tested?

A new unit test was added for checking the snapshot address. Also, this patch was tested manually in a docker cluster like so

in `/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/ozone`
```
❯ docker-compose up --scale datanode=3 -d

❯ docker exec -it ozone_om_1 bash
bash-4.2$ ozone sh volume create /vol1
bash-4.2$ ozone sh bucket create /vol1/bucket1
bash-4.2$ ozone sh key put /vol1/bucket1/key1 /etc/hosts
bash-4.2$ ozone sh key put /vol1/bucket1/dir1/key2 /etc/os-release
bash-4.2$ ozone sh snapshot create /vol1/bucket1 snap1
bash-4.2$ ozone sh key list /vol1/bucket1/.snapshot/snap1
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : ".snapshot/snap1/dir1/key2",
  "dataSize" : 393,
  "creationTime" : "2023-01-12T15:03:53.014Z",
  "modificationTime" : "2023-01-12T15:03:53.774Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  }
}, {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : ".snapshot/snap1/key1",
  "dataSize" : 174,
  "creationTime" : "2023-01-12T15:03:37.940Z",
  "modificationTime" : "2023-01-12T15:03:38.927Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  }
} ]
bash-4.2$ ozone sh key list --prefix=dir1 /vol1/bucket1/.snapshot/snap1
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : ".snapshot/snap1/dir1/key2",
  "dataSize" : 393,
  "creationTime" : "2023-01-12T15:03:53.014Z",
  "modificationTime" : "2023-01-12T15:03:53.774Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  }
} ]
bash-4.2$ ozone sh key list --prefix=.snapshot/snap1/dir1 /vol1/bucket1                
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : ".snapshot/snap1/dir1/key2",
  "dataSize" : 393,
  "creationTime" : "2023-01-12T15:03:53.014Z",
  "modificationTime" : "2023-01-12T15:03:53.774Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  }
} ]

```

xBis7 commented:
@smengcl @sadanand48 Can you please take a look at this PR?

GeorgeJahad commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotKeyHandler.java:
```diff
@@ -0,0 +1,80 @@
+    aliases = "key-ls",
+    description = "list all keys in a given " +
+        "snapshot under a bucket")
+public class ListSnapshotKeyHandler extends SnapshotHandler {
```
The original bug report was to fix the "ozone sh key list" command.  This adds a new but similar snapshot "key-list" command, which is fine.  But couldn't you also fix the "ozone sh key list" command by changing this line to be extended from "SnapshotHandler" instead of BucketHandler?: https://github.com/apache/ozone/blob/51a0311425e88b56844f05101c18a9b2075fda3a/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/ListKeyHandler.java#L42

xBis7 replied:
> But couldn't you also fix the "ozone sh key list" command by changing this line to be extended from "SnapshotHandler" instead of BucketHandler?:

@GeorgeJahad If we make that change then `ozone sh key list` will only work with snapshots and it will give us an error if we try to get the keys under a bucket.

```
bash-4.2$ ozone sh key list /vol1/bucket1
Invalid value for positional parameter at index 0 (<value>): cannot convert '/vol1/bucket1' to OzoneAddress (org.apache.hadoop.ozone.client.OzoneClientException: Snapshot name is missing.)
Usage: ozone sh key list [-hV] [-l=<limit>] [-p=<prefix>] [-s=<startItem>]
                         <value>
list all keys in a given bucket
...
```
but it works for a snapshot

```
bash-4.2$ ozone sh key list /vol1/bucket1/.snapshot/snap1
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : "key1",
  "dataSize" : 3811,
  "creationTime" : "2023-01-09T16:56:04.248Z",
  "modificationTime" : "2023-01-09T16:56:05.308Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  }
} ]
```

xBis7 commented:
@GeorgeJahad I made some changes and now it works with the `ozone sh key list` command. Could you take another look at this?

GeorgeJahad commented on hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneAddress.java:
```diff
@@ -96,14 +108,38 @@ public void checkUrlTypes() throws OzoneClientException, IOException {
+    Assert.assertEquals(".snapshot/snap1", address.getKeyName());
+
+
+    String message = "Delimiters (/) not allowed following " +
```
NIT:  if you are going to check for a precise error message, I would define it in a string constant in the class being tested, and then use that string constant in the test.  DRY!

xBis7 replied:
@GeorgeJahad It's snapshot specific and it won't be reused in another method, so I simplified it and kept only the snapshot part.

GeorgeJahad commented:
+1 LGTM

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/ListKeyHandler.java:
```diff
@@ -50,11 +51,27 @@ protected void execute(OzoneClient client, OzoneAddress address)
 
     String volumeName = address.getVolumeName();
     String bucketName = address.getBucketName();
+    String snapshotName = "";
```
```suggestion
    String snapshotName = address.getSnapshotName();
    String keyPrefix = "";

    if (!Strings.isNullOrEmpty(snapshotName)) {
      keyPrefix += snapshotName;

      if (!Strings.isNullOrEmpty(listOptions.getPrefix())) {
        keyPrefix += "/";
      }
    }

    if (!Strings.isNullOrEmpty(listOptions.getPrefix())) {
      keyPrefix += listOptions.getPrefix();
    }

hemantk-12 commented on hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneAddress.java:
```diff
@@ -47,26 +48,34 @@ public static Collection<Object[]> data() {
-    OzoneAddress address;
-
+  public void checkRootUrlType() throws OzoneClientException {
     address = new OzoneAddress("");
```
What's the difference between this and last parameter in line 47?

xBis7 replied:
There is no difference, check [here](https://github.com/apache/ozone/blob/a20e84ef03517d8970d8094cd4ec85d3d53caee2/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/OzoneAddress.java#L70), but we need this call to initialize `OzoneAddress`, even if we are setting an empty address.

hemantk-12 replied:
I still don't get to have empty string in test cases. It is unnecessary to me. But I'll leave it because it was already there.

hemantk-12 commented on hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneAddress.java:
```diff
@@ -96,14 +108,37 @@ public void checkUrlTypes() throws OzoneClientException, IOException {
+
+    address = new OzoneAddress(prefix + "vol1/bucket/.snapshot");
+
+    exception.expect(OzoneClientException.class);
+    exception.expectMessage(message);
 
+    address.ensureSnapshotAddress();
```
```suggestion
    OzoneClientException exception = assertThrows(OzoneClientException.class, () -> {
        address.ensureSnapshotAddress();
    });

    assertTrue(exception.getMessage().contains(message));
``` 

It would be nicer and simpler if you use Junit5. https://www.baeldung.com/junit-assert-exception

hemantk-12 commented on hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneAddress.java:
```diff
@@ -47,26 +48,34 @@ public static Collection<Object[]> data() {
     address.ensureRootAddress();
+  }
 
+  @Test
```
You can also upgrade these tests to Junit5.

hemantk-12 replied:
Thanks for upgrading to Junit-5.

hemantk-12 commented on hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneAddress.java:
```diff
@@ -96,14 +108,37 @@ public void checkUrlTypes() throws OzoneClientException, IOException {
 
+    address.ensureSnapshotAddress();
   }
 }
```
Please add a new line at the end.

hemantk-12 replied:
It is general practice to add new line at the end of the file. AFAIK, the idea comes from the Unix file system.
https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file, https://gist.github.com/OleksiyRudenko/d51388345ea55767b7672307fe35adf3

xBis7 replied:
@hemantk-12 That's interesting! Thanks for letting me know, I'll keep it in mind.

hemantk-12 reviewed (commented):
Thanks for the patch @xBis7

xBis7 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/ListKeyHandler.java:
```diff
@@ -21,25 +21,26 @@
-    description = "list all keys in a given bucket")
-public class ListKeyHandler extends BucketHandler {
+    description = "list all keys in a given bucket or snapshot")
+public class ListKeyHandler extends SnapshotHandler {
```
> Also structurally SnapshotHandler gives the wrong impression.

You are right, but despite the class name, it's related to what the path accepts and not the commands. We need a way to verify an address by calling `OzoneAddress.ensureSnapshotAddress()`.

hemantk-12 replied:
Even tho we need to verify address, I still don't believe that it is a good idea to have it this way if you think in term of the code structure and hierarchy.

xBis7 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/OzoneAddress.java:
```diff
@@ -343,6 +349,35 @@ public void ensureKeyAddress() throws OzoneClientException {
+      throws OzoneClientException {
+    if (keyName.length() > 0) {
+      if (OmUtils.isBucketSnapshotIndicator(keyName)) {
+        snapshotName += keyName;
```
You are right, it's misleading. I will give it a better name.

hemantk-12 reviewed (approved):
LGTM.

smengcl commented on hadoop-ozone/tools/pom.xml:
```diff
@@ -124,6 +124,10 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
     </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-params</artifactId>
+    </dependency>
```
Should this have a scope of `test`?

```suggestion
      <artifactId>junit-jupiter-params</artifactId>
      <scope>test</scope>
    </dependency>
```

xBis7 replied:
yes, IDE auto imported it and I forgot to check. I'll fix it!

smengcl commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/IncrementalContainerReportHandler.java:
```diff
@@ -88,8 +88,10 @@ public void onMessage(final IncrementalContainerReportFromDatanode report,
             id = container.containerID();
           } finally {
-            if (!replicaProto.getState().equals(
+            if (replicaProto.getState().equals(
                 ContainerReplicaProto.State.DELETED)) {
+              nodeManager.removeContainer(dd, id);
+            } else {
```
Is this change brought in accidentally by the merge conflict resolution?

https://github.com/apache/ozone/pull/4125/commits/b2bac5584d575dbf02eb1d52628a19a6034f2a81

xBis7 replied:
@smengcl Yes, and the changes in `TestIncrementalContainerReportHandler` as well.

smengcl replied:
Could be a GitHub PR diff display bug. See https://github.com/apache/ozone/pull/4134/files#r1094909600

xBis7 replied:
@smengcl What if I try another rebase?

xBis7 replied:
Another merge, fixed it.

hemantk-12 reviewed (approved):
Overall changes look good to me except my earlier comment about `ListKeyHandler ` extending `SnapshotHandler`. 

LGTM.

xBis7 commented:
@hemantk-12 Would [this approach](https://github.com/xBis7/ozone/blob/5509ccd365232933fc51c0a688a61a4e2d5d4699/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotKeyHandler.java) seem cleaner? That's from commit 5509ccd365232933fc51c0a688a61a4e2d5d4699

smengcl commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/OzoneAddress.java:
```diff
@@ -343,6 +350,35 @@ public void ensureKeyAddress() throws OzoneClientException {
+      throws OzoneClientException {
+    if (keyName.length() > 0) {
+      if (OmUtils.isBucketSnapshotIndicator(keyName)) {
+        snapshotNameWithIndicator += keyName;
```
Is there a reason to use `+=` instead of `=` here?

What if `ensureSnapshotAddress()` is called twice?

xBis7 replied:
You are right, I'll fix it!

smengcl commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotHandler.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Base class for snapshot command handlers.
+ */
+public abstract class SnapshotHandler extends Handler {
```
Should this extend `BucketHandler` instead of plain `Handler`?

xBis7 replied:
Extending `BucketHandler` won't work because we will end up calling `ensureBucketAddress()` instead of `ensureSnapshotAddress()`.

```
bash-4.2$ ozone sh key list /vol1/bucket1/.snapshot/snap1
Invalid value for positional parameter at index 0 (<value>): cannot convert '/vol1/bucket1/.snapshot/snap1' to OzoneAddress (org.apache.hadoop.ozone.client.OzoneClientException: Invalid bucket name. Delimiters (/) not allowed in bucket name)
```

smengcl replied:
Got it. Thanks @xBis7

smengcl commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/ListKeyHandler.java:
```diff
@@ -64,8 +79,18 @@ protected void execute(OzoneClient client, OzoneAddress address)
-      out().printf("Found : %d keys for bucket %s in volume : %s ",
-          counter, bucketName, volumeName);
+      if (!Strings.isNullOrEmpty(snapshotNameWithIndicator)) {
+        // snapshot[0] = .snapshot
+        // snapshot[1] = snapshot name
+        String[] snapshotValues = snapshotNameWithIndicator.split("/");
```
nit
```suggestion
        // snapshotValues[0] = ".snapshot"
        // snapshotValues[1] = snapshot name
        String[] snapshotValues = snapshotNameWithIndicator.split("/");
```

smengcl commented on hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneAddress.java:
```diff
@@ -18,25 +18,21 @@
-import org.junit.runner.RunWith;
-import org.junit.runners.Parameterized;
-import org.junit.runners.Parameterized.Parameters;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
```
Thanks for bumping this to JUnit 5!

hemantk-12 commented:
> @hemantk-12 Would [this approach](https://github.com/xBis7/ozone/blob/5509ccd365232933fc51c0a688a61a4e2d5d4699/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotKeyHandler.java) seem cleaner? That's from commit [5509ccd](https://github.com/apache/ozone/commit/5509ccd365232933fc51c0a688a61a4e2d5d4699).

TBH, I'm not sure if that it clearen either. To me, `ListKeyHandler` should be independent of `bucketPath` or `snapshotPath`. You can have different internal implementations of `ListKeyHandler` based on path type is bucket or snapshot if needed but high level it should be independent.
As I said earlier, `SnapshotHandler` says that it is to handle snapshot reqeust like snapshot create, delete or list snapshot not listing the keys from snapshot.

Again this is what I think based on my experience with Ozone so far. May be @smengcl or @prashantpogde can look and help us here.

smengcl reviewed (commented):
Thanks @xBis7 for working on this. Overall lgtm.

smengcl commented:
> @hemantk-12 Would [this approach](https://github.com/xBis7/ozone/blob/5509ccd365232933fc51c0a688a61a4e2d5d4699/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/snapshot/ListSnapshotKeyHandler.java) seem cleaner? That's from commit [5509ccd](https://github.com/apache/ozone/commit/5509ccd365232933fc51c0a688a61a4e2d5d4699).

Hi @xBis7 , looks like this approach is add a brand new subcommand to Ozone Shell? If this is the case it is not desired, as we do want a unified interface for listing under regular buckets and snapshotted ones.

On the other hand, I agree with @hemantk-12 that we should try to keep using the previous `BucketHandler` [here](https://github.com/apache/ozone/pull/4125#discussion_r1084605774). Or maybe we could rename `SnapshotHandler` into something like `BucketAndSnapshotHandler` in order not to give the wrong impression?

xBis7 commented:
@hemantk-12 @smengcl Thanks for the reviews, I've updated the patch. I get your point about the `SnapshotHandler` class.

> You can have different internal implementations of ListKeyHandler based on path type is bucket or snapshot if needed but high level it should be independent.

This would be the best, but we would have to reorganize all Handler classes and that would require changes across all shell commands. If we are to go that way, it should be in another PR.

> we could rename SnapshotHandler into something like BucketAndSnapshotHandler in order not to give the wrong impression?

This seemed the easiest approach, so I renamed the class to `BucketSnapshotHandler`.

hemantk-12 commented:
> @hemantk-12 @smengcl Thanks for the reviews, I've updated the patch. I get your point about the `SnapshotHandler` class.
> 
> > You can have different internal implementations of ListKeyHandler based on path type is bucket or snapshot if needed but high level it should be independent.
> 
> This would be the best, but we would have to reorganize all Handler classes and that would require changes across all shell commands. If we are to go that way, it should be in another PR.
> 
> > we could rename SnapshotHandler into something like BucketAndSnapshotHandler in order not to give the wrong impression?
> 
> This seemed the easiest approach, so I renamed the class to `BucketSnapshotHandler`.

Doing it as part of another PR is completely fine by me.

Thanks for change the name.

hemantk-12 commented:
LGTM. 

Feel free to merge once Siyao approves it.

smengcl commented:
Thanks @xBis7 for the patch. Thanks @GeorgeJahad @hemantk-12 for the reviews.

xBis7 commented:
@smengcl @hemantk-12 @GeorgeJahad Thanks for the reviews!
